Occlusion culling algorithm based on recursive descend

x2048 opened the pull request:
This is inspired by portal-based occlusion culling technique (https://www.youtube.com/watch?v=8xgb-ZcZV9s).

Every mapblock side is treated as a portal, and the map is walked recursively through the portals. Simple logical rule appled - a 'far' side of a mapblock is visible if :
* If adjacent 'near' sides are see-through (block is not occluded, and adjacent block does not have a solid side)
* If opposite 'near' side is see-through, and it overlaps the 'far' side (so called dominant axis in the look vector)

A side is traversed if it's visible and not fully solid, otherwise it's marked as opaque for the next mapblock.

## To do

This PR is Ready for Review.

## How to test

0. Configure server parameters to send more blocks by default:
```
max_block_send_distance = 40
max_block_generate_distance = 40
block_send_optimize_distance = 40
client_unload_unused_data_timeout = 6000
client_mapblock_limit = -1
max_simultaneous_block_sends_per_client = 80
```
1. Enter MTG and set view distance to 400-600
2. Enable F6 profiler and F5 graphs
3. Flip between `/set occlusion_culler new` and `/set occlusion_culler old`
4. Mark the change in `updateDrawList` value, `MapBlocks drawn`, `Time non-rendering` and `jitter %`
5. The game must work more smoothly even with large number of mapblocks loaded.

x2048 committed:
Mapblock culling algorithm based on recursive descend

lhofhansl commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +199,413 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+{
+public:
+	static constexpr u16 CHUNK_INNER_BITS  = 4; // number of bits to address inside a chunk
+	static constexpr u16 CHUNK_EDGE = 16; // 2 ^ CHUNK_SHIFT
```
Will this work with MAP_BLOCKSIZE > 16?

x2048 replied:
These are chunks of mapblocks. Will work with any mapblock size.

Desour commented:
Please correct me if I'm wrong, but if I understand correctly, this replaces the current occlusion culling with a more intelligent mapblock-traversal that takes occluding sides between mapblocks into account (as opposed to the current traversal which considers *all* loaded mapblocks that have a mesh). (Or actually, it does both, old and new occlusion culling.)

(This has not too much to do with the portal based occlusion culling. I.e. it doesn't shrink the view frustum.)

In general this is a reasonable thing to do.
Some things to consider:
* Mapblock meshes can exceed the mapblock bounds (the 16^3 cube). Try e.g. with the plantlike rooted thing in devtest, and set its param2 really high.
  => Too much occlusion is possible (though not likely, I guess).
* It should be sufficient to just use the +X,+Y,+Z sides of mapblocks, instead of all 6. It's probably not too common that one side is opaque and the other isn't.
* Don't traverse blocks that are far outside of the view frustum.


--------

<details><summary>out of scope stuff</summary>

Btw. to take big occluders near the player into account, we could maybe use the gpu to do occlusion testing with bounding boxes (on each frame, not draw list update):
* Take the depth-buffer of the previous frame (without the semi-transparent stuff).
* (Scale it down for less accuracy.)
* Turn off color and depth buffer write.
* For each block: Draw a bounding box. Store somewhere if anything of the mapblock is drawn. (This can probably use instanced rendering.)
* The result is the drawlist for the next frame. 
* In the next frame, also include the blocks that were partly outside of the old frustum.

</details>

lhofhansl commented:
In wire-frame mode I'm observing "flashing". Looks like updateDrawList alternates between excluding a block and not excluding it.

x2048 committed:
Comment out occlusion debugging code

x2048 commented:
This flashing shows which mapblocks are 'culled' by the algorithm. I've commented the debug code out in the latest commit.

x2048 commented:
@FossFanatic 

> What does this do exactly? Does it just make gameplay smoother?

It makes gameplay smooth at large render distances (>400) where the previous algorithm starts taking about 50ms to complete and impacts the framerate.


@Desour 

> Please correct me if I'm wrong, .. this replaces the current occlusion culling with a more intelligent mapblock-traversal that takes occluding sides between mapblocks into account ... actually, it does both, old and new occlusion culling.

Correct

> (This has not too much to do with the portal based occlusion culling. I.e. it doesn't shrink the view frustum.)

The frustum is not changed dynamically, but the logic for selecting which far sides to traverse is based on the look direction and occluded near sides, similar to how portals outside of narrowed frustum are culled in portal-based occlusion culling.

> Mapblock meshes can exceed the mapblock bounds (the 16^3 cube). Try e.g. with the plantlike rooted thing in devtest, and set its param2 really high.
    => Too much occlusion is possible (though not likely, I guess).

This is not different from the previous algorithm. Oversized mapblock meshes will be culled by normal-sized occluders and will not occlude more than normal.

> It should be sufficient to just use the +X,+Y,+Z sides of mapblocks, instead of all 6. It's probably not too common that one side is opaque and the other isn't.

There is no significant difference in storage or performance between 3 or 6 bits per mapblock.

> Don't traverse blocks that are far outside of the view frustum.

Frustum culling is there as one of the predicates of variable `culled`.

kilbith commented:
Been testing this and the new culler indeed performs better.

Old culler (viewing range: 400)

![Capture d’écran du 2022-12-30 21-19-41](https://user-images.githubusercontent.com/7883281/210109109-4e4336e3-ea74-4910-82f2-6d7adaec8a49.png)

New culler (viewing range: 400)

![Capture d’écran du 2022-12-30 21-19-28](https://user-images.githubusercontent.com/7883281/210109166-8dd8696a-945c-4ecc-8159-65c213d5ca99.png)

kilbith commented:
@FossFanatic If you don't have anything useful to say, please abstain from commenting. The pull-requests are not a place to share your feelings and your stupid questions.

lhofhansl commented:
> This flashing shows which mapblocks are 'culled' by the algorithm. I've commented the debug code out in the latest commit.

Sorry, I should do a more thorough job (doing this on the side a bit)

lhofhansl commented:
BTW. There is also server-side occlusion culling. ba4b704ebf24952ab9a84c914b8ad6c45dabfaba
This is now default on. Might want to use the same algorithm there (for that change I had move some occlusion culling code from clientmap to map among other things)

x2048 commented:
Yes, it is possible to optimize the server-side sending algorithm, both memory and performance-wise with this approach, but it is too large of a change to add to this PR.

lhofhansl commented:
Could you at least put the new culler as a method in map and call it from updateDrawList? That way it will be easier to do that on the server too (I'm happy to add the relevant server code when this is merged.)

Generally the new culler seems to be better. I was hoping it would be more accurate in long tunnels close to block boundaries, but that is not the case (but it is not worse than the original one)

Desour commented:
> Generally the new culler seems to be better. I was hoping it would be more accurate in long tunnels close to block boundaries, but that is not the case (but it is not worse than the original one)

The old occlusion test is still done, so it can only cull more.

-------------

(TBH, until now I underestimated this PR and though it would simply walk all non-opaque sides, and was therefore confused by the complex logic. But now I see how you make use of the fact that the opposite side is only visible on the dominant axis, nice. More explanation in code wouldn't hurt.)

You're currently just using a fifo queue to decide the next-to-visit block. And you store opaque sides in the flags, and consider unchecked sides (as in: the mapblock on that side wasn't visited before) as transparent, right?

I think you could cull more by only interpreting those sides as transparent, where the block on that side was visited before:

If we can see block B from block A through a side on axis C, A's C-coord is exactly 1 nearer at the camera block, and all other pos elements are the same. Therefore, the taxi-norm-distance of A to the camera block is exactly 1 smaller than that of B.
This means that there's always one current taxi-norm-distance D, and you only push blocks with t-dist D+1 to the queue. And the visits are ordered by taxi-norm.

Now, if we are in block A, and want to decide the possible 'far' sides, it is not possible that a new 'near' side would be added through a later visited block B, because B would have to have a t-dist 1 *smaller* than A.

lhofhansl commented:
Oh yeah, I now see we still call isBlockOccluded.
I had silently assumed we're doing something like this: https://tomcc.github.io/2014/08/31/visibility-1.html

So this is purely to pre-filter block on which we have to do occlusion culling. It would be great to make this good enough to remove isBlockOccluded, as that is fairly inaccurate.

x2048 committed:
Remove unused variable

x2048 committed:
Code style correction / use of constexpr

x2048 committed:
Refine the algorithm after the review

- Invert the initial assumption about visiblity of mapblock sides (now all bocks hidden unless exposed)
- More inline documentation
- Make legacy raytraced culling an option 'enable_raytraced_culling'

x2048 committed:
Remove unnecessary tests for occlusion culling

If a block is not visible from any side, it will not be added to the queue, no need for extra checks

x2048 committed:
Only run raytrace test for partially occluded blocks

This avoids repeating the test for large air volumes at high view ranges

x2048 commented:
@Desour I followed your advice, and the non-visited sides are not treated opaque. I've also added comments explaining the reasoning behind each block of code.

@lhofhansl I hadn't seen that article before, but we are doing something very close:
* breadth-first search for blocks to render
* only traverse non-opaque sides in the blocks
* side is invisible if it was never traversed by an adjacent block
* invisible sides occlude the sides behind them, also terminating the search.

I have limited the use of raytraced occlusion culling, so that obviously visible terrain is not tested and added an option for it. Unfortunately, disabling raytraced test significantly reduces the quality of culling in many scenes.

x2048 commented:
I would like to remove the old culling algorithm entirely if the new algorithm is accepted.

lhofhansl commented:
> disabling raytraced test significantly reduces the quality of culling in many scenes.

I've seen that too. In some scenes (like looking out of a tunnel) it's ~100 blocks vs ~1000 blocks drawn.
Interestingly I found FPS is not influenced that much. In a scene with view_range 600, free view, I found 1 FPS difference between the three. (14 FPS for both old, and new without raytracing, and 13 for new with ray tracing.)

Similar for the looking-out-of-tunnel scene with 10x blocks to draw with new and no raytracing. FPS is nearly the same, which I do not understand.

Perhaps I have a fast graphics card and rendering more blocks (even 10x...?!) makes no difference.
So not sure what to make of this right now.

x2048 commented:
The goal of the change is not to make a significantly better culler, but to make a faster culler. With block size of 16 and view ranges of 500+ the new algorithm is 2-3 times faster, removing much of the annoying stutter.

Block size of 32 avoids a few performance problems that I'm trying to solve here, so I'd suggest testing with size of 16.

Desour commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			// * The opposite near side (same axis) transparent, and is the dominant axis of the look vector
+
+			// dominant axis flag
+			u8 dominant_axis = (abs(look.X) > abs(look.Y) && abs(look.X) > abs(look.Z)) |
+						((abs(look.Y) > abs(look.Z) && abs(look.Y) > abs(look.Z)) << 1) |
+						((abs(look.Z) > abs(look.X) && abs(look.Z) > abs(look.Y)) << 2);
```
2nd line has typo.

Just keep this simple.
```suggestion
			u8 dominant_axis = abs(look.X) > abs(look.Y) ?
				(abs(look.X) > abs(look.Z) ? 1 : 4) :
				(abs(look.Y) > abs(look.Z) ? 2 : 4);
```

x2048 replied:
Good catch with the typo.

The dominant axis is only dominant if it's the only one with maximum value. If all three axes are equal, or if two are dominant candidates then there's no dominant axis and the far side is occluded.

Desour replied:
There's always a dominant axis if floating-point camera coordinates are used.

And if you really want to keep `look` as blockpos dir, you'd basically have multiple dominant axes, because the camera can be anywhere inside the block.

Desour commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+
+				// Test the '+' direction of the axis
+				transparency_mask <<= 1;
+				if (look[axis] >= 0 && block_coord[axis] < p_blocks_max[axis]) {
```
`p_blocks_max` is exclusive, I assume.
If `block_coord[axis] + 1 == p_blocks_max[axis]`, `next_pos` will be outside of range, and cause a buffer-overflow in `blocks_seen`.
```suggestion
				if (look[axis] >= 0 && block_coord[axis] + 1 < p_blocks_max[axis]) {
```

x2048 replied:
It's inclusive and was inclusive in the previous algorithm. blocks_seen has a + 1 in constructor to avoid overflow.

Desour commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+				if (look[axis] <= 0 && block_coord[axis] > p_blocks_min[axis]) {
+					// far side is visible if adjacent near sides are transparent, or if opposite side on dominant axis is transparent
+					bool side_visible = ((near_transparency & adjacent_sides) | (near_transparency & my_side & dominant_axis)) != 0;
+					side_visible = side_visible && ((transparency_mask & transparent_sides) != 0);
```
Why is this 2nd line needed?

x2048 replied:
If the far side is visible but fully solid, it still occludes the next block.

Desour commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+					else {
+						sides_skipped++;
+					}
+				}
```
Please make a lambda to reduce code-duplication.

x2048 replied:
Done

x2048 committed:
Update src/client/clientmap.cpp

Co-authored-by: DS <vorunbekannt75@web.de>

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			// Get the sector, block and mesh
+			MapSector *sector = this->getSectorNoGenerate(v2s16(block_coord.X, block_coord.Z));
+
+			if (!sector)
+				continue;
+
+			MapBlock *block = sector->getBlockNoCreateNoEx(block_coord.Y);
```
If the sector is not loaded, we can break early. Why run the rest of the iteration?

Desour replied:
For the same reason as for why it continues if the block or mesh doesn't exist. There might be blocks behind it.
But an early break is also fine, sure. It doesn't really hurt to treat map holes as opaque.

x2048 committed:
Optimize calculation of solid mapblock sides

* Untie solid sides information from the mesh

x2048 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1242,6 +1242,35 @@ MapBlockMesh::MapBlockMesh(MeshMakeData *data, v3s16 camera_offset):
+	for (u8 i = 0; i < 6; i++)
+		if (solid_node_counts[i] == FULL_SIDE)
+			m_solid_sides |= (1 << i);
+
```
Changed to early breaks

x2048 commented on src/mapblock.h:
```diff
@@ -469,6 +469,8 @@ class MapBlock
 	// True if we never want to cache content types for this block
 	bool do_not_cache_contents = false;
+	// marks the sides which are opaque: 00+Z-Z+Y-Y+X-X
+	u8 solid_sides {0};
```
I've removed m_solid_sides from the mesh, it's not only in mapblock.

x2048 commented on src/client/game.cpp:
```diff
@@ -696,6 +696,7 @@ struct GameRunData {
 
 	float damage_flash;
 	float update_draw_list_timer;
+	float touch_blocks_timer;
```
[This line](https://github.com/minetest/minetest/blob/master/src/client/game.cpp#L1124) initializes GameRunData with zeros.

x2048 committed:
Remove unnecessary includes

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -28,9 +28,12 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
+#include <algorithm>
+#include <queue>
+#include <bitset>
```
Queue's used, removed the others.

x2048 committed:
Rename onSettingsChanged->on_settings_changed

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -60,6 +63,10 @@ void MeshBufListList::add(scene::IMeshBuffer *buf, v3s16 position, u8 layer)
 	list.emplace_back(l);
 }
 
+static void onSettingsChanged(const std::string &name, void *data)
```
Fixed.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
-	m_drawlist.clear();
+}
+
+void ClientMap::touchMapBlocks()
```
This was the side effect of the previous algorithm to keep loaded blocks in range warm to avoid reloads from server. I only kept existing functionality.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			// do not test sides where we are 'inside' the block's coordinates
+			u8 ignore_inner_sides = (look.X == 0 ? 3 : 0) |
+				(look.Y == 0 ? 12 : 0) |
+				(look.Z == 0 ? 48 : 0);
```
Yes, when moving along an axis of the camera block, we enter the next mapblock from one side, and all the other 5 are not 'near' and must not occlude even if not transparent.

Desour replied:
Ah, I see how this makes sense for the camera block.

However, if we take, for example, the blocks that are in the same Y-plane and Z-plane (so, the blocks in X-direction), we can only visit them from X-direction. Now, if the X side is visible from outside, but blocked from inside, it should not be transparent (but `ignore_inner_sides` makes it transparent).

I suggest to instead set `near_transparency` to 0x3F iff it's the camera block.

x2048 replied:
You are right, I changed it in the latest commit, so that only the camera block's sides are treated transparent (necessary to kick-start the algorithm), but for the other in-plane blocks non-dominant sides are treated unconditionally opaque, and it removes more blocks along the planes of the player/camera without raytraced culling.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			u8 transparent_sides = (occlusion_culling_enabled && block) ? ~block->solid_sides : 0x3F;
+
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
```
The sides are ignored because they are not 'near', as described earlier. Occlusion rules must be ignored, as if they are transparent. This only affects blocks in the same planes or on the same axes as the camera.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+		return v3s16((pos .X - base.X) >> CHUNK_INNER_BITS, (pos.Y - base.Y) >> CHUNK_INNER_BITS, (pos.Z - base.Z) >> CHUNK_INNER_BITS);
+	}
+
+	std::vector<std::unique_ptr<Chunk>> chunks;
```
No, at view range of 600 (my target), a sparse array uses around 10-15% of memory a vector would use, because vast majority of chunks would be frustum-culled. But a sidelength of 16 blocks is indeed too large for smaller distances, I'm replacing it with 8 (has just 10% overhead at view range of 100).

x2048 committed:
Optimize MapBlockFlags.

* Reduce the number of operations per access by consolidating all bit-shifts
* Reduce the chunk size to 8x8x8 from 16x16x16 mapblocks

x2048 committed:
Simplify frustum culling for blocks without meshes

Use fixed mesh radius of 0.0 if the mesh is not present.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			else {
+				mesh_sphere_center = intToFloat(block_pos_nodes, BS) + v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+				mesh_sphere_radius = std::sqrt(3.) * MAP_BLOCKSIZE * BS / 2;
+			}
```
Radius of 0.0 is good, but I would not do max/min, let the mesh decide it.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+
+			// Calculate the vector from the camera block to the current block
+			// We use it to determine through which sides of the current block we can continue the search
+			v3s16 look = block_coord - camera_block;
```
Yes, but only for the dominant axis. In all other cases comparison with 0 suddenly becomes a complex equation.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
+
+			near_transparency = (near_transparency | (near_transparency >> 1)) & 0x15;
```
The mask is not needed, because we hand-pick the bits later.

x2048 committed:
Small fixes

* Fix a typo in dominant axis calculation
* Remove redundant mask in bit operations

x2048 committed:
Fix grammar in a comment

Co-authored-by: DS <vorunbekannt75@web.de>

x2048 committed:
Use precise look direction to calculate dominant axis

lhofhansl commented:
> so I'd suggest testing with size of 16.

I did that test above with MAP_BLOCKSIZE = 16.

x2048 committed:
Code cleanup

* Deduplicate side traversal code
* Better name for far side mask flag

x2048 committed:
Fix typo when detecting solid block sides

lhofhansl commented:
```
[New Thread 0x7fff8e1e2640 (LWP 23340)]
2023-01-02 16:47:13: ACTION[Server]: singleplayer [127.0.0.1] joins game. List of players: singleplayer

Thread 1 "minetest" received signal SIGSEGV, Segmentation fault.
0x00000000004f708a in ClientMap::updateDrawList() ()
Missing separate debuginfos, use: dnf debuginfo-install bzip2-libs-1.0.8-11.fc36.x86_64 cyrus-sasl-lib-2.1.27-18.fc36.x86_64 dbus-libs-1.14.4-1.fc36.x86_64 elfutils-libelf-0.188-3.fc36.x86_64 expat-2.5.0-1.fc36.x86_64 flac-libs-1.3.4-1.fc36.x86_64 freetype-2.12.1-2.fc36.x86_64 glib2-2.72.3-1.fc36.x86_64 glibc-2.35-20.fc36.x86_64 gmp-6.2.1-2.fc36.x86_64 graphite2-1.3.14-9.fc36.x86_64 gsm-1.0.19-7.fc36.x86_64 harfbuzz-4.0.0-2.fc36.x86_64 jsoncpp-1.9.5-2.fc36.x86_64 krb5-libs-1.19.2-12.fc36.x86_64 lame-libs-3.100-12.fc36.x86_64 leveldb-1.23-4.fc36.x86_64 libICE-1.0.10-8.fc36.x86_64 libSM-1.2.3-10.fc36.x86_64 libX11-1.7.3.1-2.fc36.x86_64 libX11-xcb-1.7.3.1-2.fc36.x86_64 libXcursor-1.2.0-7.fc36.x86_64 libXext-1.3.4-8.fc36.x86_64 libXfixes-6.0.0-3.fc36.x86_64 libXi-1.8-2.fc36.x86_64 libXrender-0.9.10-16.fc36.x86_64 libXxf86vm-1.1.4-18.fc36.x86_64 libasyncns-0.8-22.fc36.x86_64 libbrotli-1.0.9-7.fc36.x86_64 libcom_err-1.46.5-2.fc36.x86_64 libcurl-7.82.0-12.fc36.x86_64 libdrm-2.4.114-1.fc36.x86_64 libevent-2.1.12-6.fc36.x86_64 libgcc-12.2.1-4.fc36.x86_64 libgcrypt-1.10.1-3.fc36.x86_64 libglvnd-1.4.0-4.fc36.x86_64 libidn2-2.3.4-1.fc36.x86_64 libjpeg-turbo-2.1.2-2.fc36.x86_64 libnghttp2-1.51.0-1.fc36.x86_64 libogg-1.3.5-3.fc36.x86_64 libpciaccess-0.16-6.fc36.x86_64 libpng-1.6.37-12.fc36.x86_64 libpsl-0.21.1-5.fc36.x86_64 libselinux-3.3-4.fc36.x86_64 libsndfile-1.1.0-4.fc36.x86_64 libssh-0.9.6-4.fc36.x86_64 libstdc++-12.2.1-4.fc36.x86_64 libunistring-1.0-1.fc36.x86_64 libuuid-2.38-1.fc36.x86_64 libvorbis-1.3.7-5.fc36.x86_64 libxcb-1.13.1-9.fc36.x86_64 libxcrypt-4.4.33-4.fc36.x86_64 libzstd-1.5.2-2.fc36.x86_64 llvm-libs-14.0.5-1.fc36.x86_64 luajit-2.1.0-0.25beta3.fc36.x86_64 lz4-libs-1.9.3-4.fc36.x86_64 mesa-libGL-22.1.7-1.fc36.x86_64 mesa-libGLU-9.0.1-6.fc36.x86_64 mpg123-libs-1.31.1-1.fc36.x86_64 ncurses-libs-6.2-9.20210508.fc36.x86_64 nss-mdns-0.15.1-5.fc36.x86_64 openal-soft-1.21.1-2.fc36.x86_64 openldap-2.6.3-1.fc36.x86_64 openssl-libs-3.0.5-2.fc36.x86_64 opus-1.3.1-10.fc36.x86_64 pcre-8.45-1.fc36.1.x86_64 pipewire-jack-audio-connection-kit-0.3.63-1.fc36.x86_64 pipewire-libs-0.3.63-1.fc36.x86_64 snappy-1.1.9-4.fc36.x86_64 spatialindex-1.9.3-7.fc36.x86_64 sqlite-libs-3.36.0-5.fc36.x86_64 xz-libs-5.2.5-9.fc36.x86_64 zlib-1.2.11-33.fc36.x86_64
(gdb) bt
#0  0x00000000004f708a in ClientMap::updateDrawList() ()
#1  0x000000000054abb9 in Game::updateFrame(ProfilerGraph*, RunStats*, float, CameraOrientation const&) ()
#2  0x000000000054eb4a in Game::run() ()
#3  0x000000000054f7ec in the_game(bool*, InputHandler*, RenderingEngine*, GameStartData const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, ChatBackend&, bool*) ()
#4  0x00000000004ed97f in ClientLauncher::run(GameStartData&, Settings const&) ()
#5  0x00000000004bf8d0 in main ()
```

Sorry. Don't have a debug build handy.

x2048 committed:
Fix crash in MapBlockFlags and simplify the code

x2048 commented:
The crash is fixed.

kilbith commented:
Tested with latest changes, still some glitches: https://www.youtube.com/watch?v=GHa0d2IUQAg

Desour commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +202,403 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
+
+			near_transparency |= (near_transparency >> 1);
```
https://github.com/minetest/minetest/pull/13104#discussion_r1060213082
> The mask is not needed, because we hand-pick the bits later.

If you don't use the mask, you'll shift the bits into different axes.

Isn't what you wanted to do here `X = -X || +X` via `-X = -X || +X; X = -X`?

```suggestion
			near_transparency |= (near_transparency >> 1) & 0x3F;
```

x2048 replied:
The following lines hand-pick bits 0x01, 0x04 and 0x10 (negative sides of the three axes), which together are the same as the mask 0x15 that I removed. It does not matter if the other bits are pollutedm as they are ignored.

Desour replied:
Ah I see now, ok.

<!---
The important part is that the AND happens *before* the OR. (Your old code did this wrong.)

Example:
Given: `near_transparency = 0b000100 = 0x4` (-Y only)
Resulting `near_transparency` should be `0b010`, right?
Without mask: after this line: `near_transparency = 0b000110` (-Y | +X). After hand-picking: 
-->

x2048 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1582,3 +1582,31 @@ video::SColor encode_light(u16 light, u8 emissive_light)
+	v3s16 blockpos_nodes = data->m_blockpos * MAP_BLOCKSIZE;
+	const NodeDefManager *ndef = data->m_client->ndef();
+
+	u8 result = 0x3F; // all sides solid;
+
+	for (s16 i = 0; i < MAP_BLOCKSIZE && result != 0; i++)
+	for (s16 j = 0; j < MAP_BLOCKSIZE && result != 0; j++) {
+		v3s16 positions[6] = {
+			v3s16(0, i, j),
+			v3s16(MAP_BLOCKSIZE - 1, i, j),
+			v3s16(i, 0, j),
+			v3s16(i, MAP_BLOCKSIZE - 1, j),
+			v3s16(i, j, 0),
+			v3s16(i, j, MAP_BLOCKSIZE - 1)
+		};
+
+		for (u8 k = 0; k < 6; k++) {
+			const MapNode &top = data->m_vmanip.getNodeRefUnsafe(blockpos_nodes + positions[k]);
+			if (ndef->get(top).solidness != 2)
+				result &= ~(1 << k);
+		}
+	}
+
+	return result;
```
This code checks the entire cube, not the sides.

Desour replied:
The x-sidelength of the axis-aligned box with corners `v3s16(0, 0, 0), v3s16(1, B, B)` is `1-0 = 1`. This is the x-side.

(There was a mistake in my suggestion (volume=0), fixed that.)

x2048 replied:
I am not sure the suggestion is more optimal or is easier to read. I blended the loops exactly to avoid the overhead of repeating for-for loops 6 times, and the inner loop of 6 iterations is easy for compiler to unroll.

Reflecting on what people mostly play, the early breaks may not be a significant gain, because 40-99% of the blocks around the player are solid or have mostly full solid sides, unless the player is flying high in the skies.

I can change the (result != 0) condition into something more readable if that's a problem.

Desour replied:
Fine then. This code is also not that performance critical anyway, I guess.

x2048 committed:
Update src/client/clientmap.cpp

Co-authored-by: DS <vorunbekannt75@web.de>

x2048 committed:
Fix calculation of precise_look

kilbith commented:
> Tested with latest changes, still some glitches: https://www.youtube.com/watch?v=GHa0d2IUQAg

Last commit fixes it.

x2048 committed:
Treat unvisited sides of axis-aligned blocks as opaque

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -180,17 +202,403 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+	static constexpr std::size_t CHUNK_VOLUME = CHUNK_EDGE * CHUNK_EDGE * CHUNK_EDGE; // volume of a chunk
+
+	MapBlockFlags(v3s16 min_pos, v3s16 max_pos)
+			: min_pos(min_pos), volume((max_pos - min_pos) / CHUNK_EDGE + 1)
```
No, it crashes with your proposal.

lhofhansl commented:
Looks better and better. The raytraced culler is still needed, but less and less so.

I'd really love if you moved the code into Map (instead of ClientMap), but that we can do later.

+1

Desour commented on src/client/clientmap.cpp:
```diff
@@ -400,10 +401,15 @@ void ClientMap::updateDrawList()
 			// This bitset is +Z-Z+Y-Y+X-X (See MapBlockMesh), and axis is XYZ.
-			u8 transparent_sides = (occlusion_culling_enabled && block) ? ~block->solid_sides : 0x3F;
+			// Get he block's transparent sides
+			u8 transparent_sides = (occlusion_culling_enabled && block && block_inner_sides != 0x3F) ? ~block->solid_sides : 0x3F;
```
Now the camera block can't have opaque far sides.

x2048 replied:
Indeed, good catch. Fixed now, thank you!

Desour commented:
Tested again.
Found no (new) visual bugs.
Ray culling should stay on for now. It can be very effective, i.e. in near-surface caves.
The option to disable the new culler seems useless to me, and it causes too much code duplication imo.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -623,6 +623,13 @@ update_last_checked (Last update check) string
 update_last_known (Last known version update) int 0
 
+#    Type of occlusion_culler
+occlusion_culler (Occlusion Culler) enum old old,new
```
Fewer settings the better, can we delete the old culler?

x2048 replied:
We will delete it in a month.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -623,6 +623,13 @@ update_last_checked (Last update check) string
+
+#    Use raytraced occlusion culling in the new culler.
+#	 This flag enables use of raytraced occlusion culling test
+enable_raytraced_culling (Enable Raytraced Culling) bool true
```
What are the benefits and disadvantages of this? What are the usecases of disabling it?

lhofhansl replied:
Without the extra (old) raytraced culler there are fewer blocks filtered. In some scenes that can be dramatic (like 10x as many blocks). We can not yet delete the old culler.

I support the extra setting, though, as eventually we would want to remove the old raytraced culler. 
One thing I am keen about is to not let the server and clients cullers drift too far apart. Hence my desire to put this into Map (instead of clientmap) and it this to the server culler as well.

x2048 replied:
This setting allows for testing more of the new culler while in beta-testing. We can remove the setting when removing the old traversing algorithm.

x2048 commented:
@Desour The plan is to merge with two traverse algorithms. having the new one as the default, to give some more beta testing. If there are no problems we will remove the old traverse algorithm in a month and start working on the server-side.

x2048 committed:
Change default culler to bfs + documentation

x2048 committed:
Decouple camera block shortcut from side transparency.

Improves performance when camera block has solid sides (e.g. standing near a wall in a cave)
